metagpt/configs/models_config.py

```diff
@@ -0,0 +1,112 @@
+from metagpt.const import CONFIG_ROOT, METAGPT_ROOT
+from metagpt.utils.yaml_model import YamlModel
+
+
+class ModelsConfig(YamlModel):
+    """
+    Configuration class for `models` in `config2.yaml`.
+
+    Attributes:
+        models (Dict[str, LLMConfig]): Dictionary mapping model names or types to LLMConfig objects.
+
+    Methods:
+        update_llm_model(cls, value): Validates and updates LLM model configurations.
+        from_home(cls, path): Loads configuration from ~/.metagpt/config2.yaml.
+        default(cls): Loads default configuration from predefined paths.
+        get(self, name_or_type: str) -> Optional[LLMConfig]: Retrieves LLMConfig by name or API type.
+    """
+
+    models: Dict[str, LLMConfig] = Field(default_factory=dict)
+
+    @field_validator("models", mode="before")
+    @classmethod
+    def update_llm_model(cls, value):
+        """
+        Validates and updates LLM model configurations.
+
+        Args:
+            value (Dict[str, Union[LLMConfig, dict]]): Dictionary of LLM configurations.
+
+        Returns:
+            Dict[str, Union[LLMConfig, dict]]: Updated dictionary of LLM configurations.
+        """
+        for key, config in value.items():
+            if isinstance(config, LLMConfig):
+                config.model = config.model or key
+            elif isinstance(config, dict):
+                config["model"] = config.get("model") or key
+        return value
+
+    @classmethod
+    def from_home(cls, path):
+        """
+        Loads configuration from ~/.metagpt/config2.yaml.
+
+        Args:
+            path (str): Relative path to configuration file.
+
+        Returns:
+            Optional[ModelsConfig]: Loaded ModelsConfig object or None if file doesn't exist.
+        """
+        pathname = CONFIG_ROOT / path
+        if not pathname.exists():
+            return None
+        return ModelsConfig.from_yaml_file(pathname)
+
+    @classmethod
+    def default(cls):
+        """
+        Loads default configuration from predefined paths.
+
+        Returns:
+            ModelsConfig: Default ModelsConfig object.
+        """
+        default_config_paths: List[Path] = [
+            METAGPT_ROOT / "config/config2.yaml",
+            CONFIG_ROOT / "config2.yaml",
+        ]
+
+        dicts = [ModelsConfig.read_yaml(path) for path in default_config_paths]
+        final = merge_dict(dicts)
+        return ModelsConfig(**final)
+
+    def get(self, name_or_type: str) -> Optional[LLMConfig]:
+        """
+        Retrieves LLMConfig object by name or API type.
+
+        Args:
+            name_or_type (str): Name or API type of the LLM model.
+
+        Returns:
+            Optional[LLMConfig]: LLMConfig object if found, otherwise None.
+        """
+        if not name_or_type:
+            return None
+        model = self.models.get(name_or_type)
+        if model:
+            return model
+        for m in self.models.values():
+            if m.api_type == name_or_type:
```

Comment:
If there are multi model_name1 model_name2 which both use api_type=openai. Here will return the first one if name_or_type="openai", whether it meets expectation?

Reply:
```
models:
#  "YOUR_MODEL_NAME_1 or YOUR_API_TYPE_1": # model: "gpt-4-turbo"  # or gpt-3.5-turbo
#    api_type: "openai"  # or azure / ollama / groq etc.
#    base_url: "YOUR_BASE_URL"
#    api_key: "YOUR_API_KEY"
#    proxy: "YOUR_PROXY"  # for LLM API requests
#    # timeout: 600 # Optional. If set to 0, default value is 300.
#    # Details: https://azure.microsoft.com/en-us/pricing/details/cognitive-services/openai-service/
#    pricing_plan: "" # Optional. Use for Azure LLM when its model name is not the same as OpenAI's
#  "YOUR_MODEL_NAME_2 or YOUR_API_TYPE_2": # api_type: "openai"  # or azure / ollama / groq etc.
#    api_type: "openai"  # or azure / ollama / groq etc.
#    base_url: "YOUR_BASE_URL"
#    api_key: "YOUR_API_KEY"
#    proxy: "YOUR_PROXY"  # for LLM API requests
#    # timeout: 600 # Optional. If set to 0, default value is 300.
#    # Details: https://azure.microsoft.com/en-us/pricing/details/cognitive-services/openai-service/
#    pricing_plan: "" # Optional. Use for Azure LLM when its model name is not the same as OpenAI's
```
As expected.

The configuration is in the form of a map, so the user should be aware that keys must be unique. If fetching without specifying a key, it means any match will suffice.